cmd/gen-go-sample/main_test.go

```diff
@@ -299,6 +347,61 @@ func initTestGenerator() *generator {
 	}
 }
 
+func labelp(l descriptor.FieldDescriptorProto_Label) *descriptor.FieldDescriptorProto_Label {
+	return &l
+}
+
+// mapTypeAndField creates the generated MapEntry protobuf message and the actually map field.
+// If valueType is enum or message, vTypeName must not be empty.
+func mapTypeAndField(
+	fieldName string,
+	parentTyp string,
+	keyType *descriptor.FieldDescriptorProto_Type,
+	valueType *descriptor.FieldDescriptorProto_Type,
+	vTypeName string,
+) (*descriptor.DescriptorProto, *descriptor.FieldDescriptorProto) {
+
+	var vf *descriptor.FieldDescriptorProto
+	if vTypeName == "" {
+		if *valueType == descriptor.FieldDescriptorProto_TYPE_ENUM {
+			panic(errors.E(nil, "expecting non-empty enum type name"))
+		}
+		if *valueType == descriptor.FieldDescriptorProto_TYPE_MESSAGE {
+			panic(errors.E(nil, "expecting non-empty message type name"))
+		}
+		vf = &descriptor.FieldDescriptorProto{
+			Name: proto.String("value"),
+			Type: valueType,
+		}
+	} else {
+		vf = &descriptor.FieldDescriptorProto{
+			Name:     proto.String("value"),
+			Type:     valueType,
+			TypeName: proto.String(vTypeName),
+		}
+	}
+
+	mapEntry := &descriptor.DescriptorProto{
+		Name: proto.String("MapFieldEntry"),
+		Field: []*descriptor.FieldDescriptorProto{
+			{Name: proto.String("key"), Type: keyType},
+			vf,
+		},
+		Options: &descriptor.MessageOptions{
+			MapEntry: proto.Bool(true),
+		},
+	}
+
+	mf := &descriptor.FieldDescriptorProto{
```

Comment:
nit: suggest s/`mf`/`mapField`/ for symmetry with `mapEntry`and because I have a bias against too-short names